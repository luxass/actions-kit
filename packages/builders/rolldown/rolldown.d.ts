declare module "@actions-kit/config" {
	import type { RolldownOptions } from "rolldown";

	interface ActionsKitConfig {
		rolldown?: RolldownOptions;
	}
}

export {};
