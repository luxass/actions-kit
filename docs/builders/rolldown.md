# Rolldown Builder

Build GitHub Actions using Rolldown.

## Installation

::: code-group
  ```bash [pnpm]
  pnpm add -D actions-kit @actions-sdk/rolldown-builder
  ```

  ```bash [yarn]
  yarn add -D actions-kit @actions-sdk/rolldown-builder
  ```

  ```bash [npm]
  npm install -D actions-kit @actions-sdk/rolldown-builder
  ```

  ```bash [bun]
  bun install -D actions-kit @actions-sdk/rolldown-builder
  ```
:::
