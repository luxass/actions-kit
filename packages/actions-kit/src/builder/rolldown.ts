import { inferModuleType, inferOutputFilename } from "../utils";
import type { Config } from "../config";
import actionsKit from "unplugin-actions-kit/rolldown";
import { join } from "node:path";
import { builtinModules } from "node:module";

export async function build(config: Config) {
	const rolldown = await import("rolldown").then((m) => m.rolldown);
	const outputFileName = await inferOutputFilename(config);
	const libraryType = await inferModuleType(config, outputFileName);

	const rolldownConfig = config.rolldown;

	const bundle = await rolldown({
		input: "./src/index.ts",
		resolve: {
			conditionNames: ["import"],
		},
    platform: "node",
		external: [...builtinModules, ...builtinModules.map((m) => `node:${m}`)],
		plugins: [
			actionsKit({
				// TODO: allow users to specify it.
				actionPath: join(process.cwd(), "./action.yml"),
				inject: true,
			}),
		],
	});

	const wrote = await bundle.write({
		format: libraryType,
		sourcemap: false,
	});

  console.log(wrote);
}
