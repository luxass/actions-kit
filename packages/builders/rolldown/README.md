# @actions-kit/rolldown-builder

Rolldown builder for Actions Kit.
