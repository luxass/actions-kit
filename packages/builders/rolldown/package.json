{
  "name": "@actions-sdk/rolldown-builder",
  "version": "0.2.7",
  "description": "Rolldown builder for Actions Kit",
  "type": "module",
  "author": {
    "name": "Lucas Nørgård",
    "email": "lucasnrgaard@gmail.com",
    "url": "https://luxass.dev"
  },
  "packageManager": "pnpm@9.15.6",
  "license": "MIT",
  "homepage": "https://github.com/luxass/actions-kit",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/luxass/actions-kit.git",
    "directory": "packages/builders/rolldown"
  },
  "bugs": {
    "url": "https://github.com/luxass/actions-kit/issues"
  },
  "keywords": [
    "rolldown",
    "actions-kit",
    "github actions",
    "actions"
  ],
  "exports": {
    ".": {
      "import": {
        "types": "./dist/index.d.ts",
        "default": "./dist/index.mjs"
      },
      "require": {
        "types": "./dist/index.d.cts",
        "default": "./dist/index.cjs"
      }
    },
    "./package.json": "./package.json"
  },
  "main": "dist/index.cjs",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsup --tsconfig tsconfig.build.json",
    "dev": "tsup --watch",
    "typecheck": "tsc --noEmit"
  },
  "peerDependencies": {
    "rolldown": "catalog:",
    "actions-kit": "workspace:*"
  },
  "dependencies": {
    "defu": "catalog:",
    "unplugin-actions-kit": "workspace:*"
  },
  "devDependencies": {
    "@types/node": "catalog:",
    "actions-kit": "workspace:*",
    "rolldown": "catalog:",
    "tsup": "catalog:",
    "typescript": "catalog:",
    "vitest": "catalog:"
  }
}
