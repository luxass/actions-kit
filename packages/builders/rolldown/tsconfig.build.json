{
  "extends": ["../../../tsconfig.base.build.json"],
  "include": ["rolldown.d.ts", "src/**/*.ts"]
}
