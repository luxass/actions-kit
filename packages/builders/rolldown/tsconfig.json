{
  "extends": ["../../../tsconfig.base.json"],
  "include": ["rolldown.d.ts", "src/**/*.ts"]
}
