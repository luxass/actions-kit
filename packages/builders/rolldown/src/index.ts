import { stat } from "node:fs/promises";
import { builtinModules } from "node:module";
import { join } from "node:path";
import { type BuildOutput, defineBuilder } from "actions-kit/builder";
import { inferModuleType, inferOutputFilename } from "actions-kit/builder-utils";
import { defu } from "defu";
import type { RolldownOptions } from "rolldown";
import { rolldown } from "rolldown";
import RolldownActionsKit from "unplugin-actions-kit/rolldown";

export default function rolldownBuilder(options: RolldownOptions = {}) {
	return defineBuilder({
		name: "rolldown",
		build: async ({ cwd, config }) => {
			const outputFileName = await inferOutputFilename(config);
			const libraryType = await inferModuleType(config, outputFileName);
			const rolldownOptions = defu(options, {
				input: config.build?.input || join(cwd, "src/index.ts"),
				external: [...builtinModules, ...builtinModules.map((m) => `node:${m}`)],
				resolve: {
					conditionNames: ["import"],
				},
				platform: "node",
				plugins: [
					RolldownActionsKit({
						actionPath: join(cwd, "./action.yml"),
						inject: config.inject,
						autocomplete: config.autocomplete,
					}),
				],
				output: {
					name: outputFileName,
					format: libraryType,
				},
				onwarn: (warning) => {
					if (
						warning.code === "UNRESOLVED_IMPORT" ||
						warning.code === "CIRCULAR_DEPENDENCY" ||
						warning.code === "EMPTY_BUNDLE"
					) {
						return;
					}

					// TODO: pretty print warnings
				},
			} satisfies RolldownOptions);

			const builder = await rolldown(rolldownOptions);

			const output = await builder.write({
				format: libraryType,
				sourcemap: false,
				dir: join(cwd, "dist"),
				entryFileNames: outputFileName,
			});

			if (Array.isArray(output) && output.length > 1) {
				console.warn("Multiple outputs detected. Only the first output will be displayed.");
			}

			const result = Array.isArray(output) ? output[0] : output;

			const outputs: BuildOutput[] = [];
			for (const _result of result.output) {
				const stats = await stat(join(cwd, "dist", _result.fileName));
				outputs.push({
					name: _result.fileName,
					path: join(cwd, "dist", _result.fileName),
					size: stats.size,
				});
			}

			return outputs;
		},
	});
}
