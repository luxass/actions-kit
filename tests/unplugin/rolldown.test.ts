import { readFile } from "node:fs/promises";
import { join } from "node:path";
import { rolldown } from "rolldown";
import { expect, it } from "vitest";
import { fromFileSystem, testdir } from "vitest-testdirs";
import ActionKitPlugin from "../../src/unplugin/rolldown";

it("expect `actions-kit.d.ts` to be generated", async () => {
  const directoryJson = await fromFileSystem("./tests/fixtures/action.yaml");
  const testdirPath = await testdir(directoryJson);

  expect(testdirPath).toBeDefined();

  const bundle = await rolldown({
    input: join(testdirPath, "index.ts"),
    platform: "node",
    external: ["@actions/core"],
    plugins: [
      ActionKitPlugin({
        actionPath: join(testdirPath, "action.yaml"),
      }),
    ],
  });

  const { output } = await bundle.generate({
    format: "cjs",
    sourcemap: false,
  });

  const dtsOutput = await readFile(join(testdirPath, "actions-kit.d.ts"), "utf-8");

  expect(output[0]).toBeDefined();
  expect(output[0].code).toBeDefined();

  expect(output[0].code).toMatchSnapshot();
  expect(dtsOutput).toMatchSnapshot();
});

it("expect `actions-kit.d.ts` to include `ACTION_INPUTS`", async () => {
  const directoryJson = await fromFileSystem("./tests/fixtures/action.yaml");
  const testdirPath = await testdir(directoryJson);

  expect(testdirPath).toBeDefined();

  const bundle = await rolldown({
    input: join(testdirPath, "index.ts"),
    platform: "node",
    external: ["@actions/core"],
    plugins: [
      ActionKitPlugin({
        actionPath: join(testdirPath, "action.yaml"),
        injectInputs: true,
      }),
    ],
  });

  const { output } = await bundle.generate({
    format: "cjs",
    sourcemap: false,
  });

  const dtsOutput = await readFile(join(testdirPath, "actions-kit.d.ts"), "utf-8");

  expect(output[0]).toBeDefined();
  expect(output[0].code).toBeDefined();
  expect(output[0].code).toContain("ACTION_INPUTS");
  expect(dtsOutput).toContain("ACTION_INPUTS");

  expect(output[0].code).toMatchSnapshot();
  expect(dtsOutput).toMatchSnapshot();
});
