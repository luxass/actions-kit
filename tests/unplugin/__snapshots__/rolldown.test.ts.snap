// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`expect \`actions-kit.d.ts\` to be generated 1`] = `
""use strict";
//#region rolldown:runtime
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
		key = keys[i];
		if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
			get: ((k) => from[k]).bind(null, key),
			enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
		});
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));

//#endregion
const core = __toESM(require("@actions/core"));

//#region .vitest-testdirs/vitest-rolldown-expect-actions-kit-d-ts-to-be-generated/index.ts
async function run() {
	const type = core.getInput("type");
	const whoToGreet = core.getInput("who-to-greet");
	const message = \`Hello \${whoToGreet}!\`;
	core.info(message);
	if (type === "error") {
		core.setFailed("This is an error!");
	}
	core.setOutput("message", message);
}
run().catch((err) => {
	console.error(err);
	core.setFailed(err.message);
});

//#endregion"
`;

exports[`expect \`actions-kit.d.ts\` to be generated 2`] = `
"// generated by 'actions-kit'
        import "@actions/core";

        

declare module "@actions/core" {

  type InputNames = "type" | "who-to-greet";

  export function getInput(name: InputNames, options?: InputOptions): string;
}
"
`;

exports[`expect \`actions-kit.d.ts\` to include \`ACTION_INPUTS\` 1`] = `
""use strict";
//#region rolldown:runtime
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
		key = keys[i];
		if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
			get: ((k) => from[k]).bind(null, key),
			enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
		});
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));

//#endregion
const core = __toESM(require("@actions/core"));

//#region .vitest-testdirs/vitest-rolldown-expect-actions-kit-d-ts-to-include-ACTION_INPUTS/index.ts
globalThis.ACTION_INPUTS = {
	"type": "type",
	"who-to-greet": "who-to-greet"
};
async function run() {
	const type = core.getInput("type");
	const whoToGreet = core.getInput("who-to-greet");
	const message = \`Hello \${whoToGreet}!\`;
	core.info(message);
	if (type === "error") {
		core.setFailed("This is an error!");
	}
	core.setOutput("message", message);
}
run().catch((err) => {
	console.error(err);
	core.setFailed(err.message);
});

//#endregion"
`;

exports[`expect \`actions-kit.d.ts\` to include \`ACTION_INPUTS\` 2`] = `
"// generated by 'actions-kit'
        import "@actions/core";

        
declare global {
  export const ACTION_INPUTS: {
  "type": "type",
  "who-to-greet": "who-to-greet"
};
}

declare module "@actions/core" {

  type InputNames = "type" | "who-to-greet";

  export function getInput(name: InputNames, options?: InputOptions): string;
}
"
`;
